src/pydvl/utils/caching/config.py

```diff
@@ -0,0 +1,40 @@
+from dataclasses import dataclass, field
+from typing import Collection
+
+__all__ = ["CachedFuncConfig"]
+
+
+@dataclass
+class CachedFuncConfig:
+    """Configuration for cached functions and methods, providing
+    memoization of function calls.
+
+    Instances of this class are typically used as arguments for the construction
+    of a [Utility][pydvl.utils.utility.Utility].
+
+    Args:
+        ignore_args: Do not take these keyword arguments into account when
+            hashing the wrapped function for usage as key. This allows
+            sharing the cache among different jobs for the same experiment run if
+            the callable happens to have "nuisance" parameters like `job_id` which
+            do not affect the result of the computation.
+        time_threshold: Computations taking less time than this many seconds are
+            not cached. A value of 0 means that it will always cache results.
+        allow_repeated_evaluations: If `True`, repeated calls to a function
+            with the same arguments will be allowed and outputs averaged until the
+            running standard deviation of the mean stabilizes below
+            `rtol_stderr * mean`.
+        rtol_stderr: relative tolerance for repeated evaluations. More precisely,
+            [memcached()][pydvl.utils.caching.memcached] will stop evaluating the function once the
+            standard deviation of the mean is smaller than `rtol_stderr * mean`.
+        min_repetitions: minimum number of times that a function evaluation
+            on the same arguments is repeated before returning cached values. Useful
+            for stochastic functions only. If the model training is very noisy, set
+            this number to higher values to reduce variance.
+    """
+
+    ignore_args: Collection[str] = field(default_factory=list)
+    time_threshold: float = 0
```

Comment:
But time thresholds should depend on the latency of the caching backend, shouldn't they? We definitely don't want to cache every function call to memcached and definitely not to disk, so that a 0 here is not good for the users.

The default 0.3 was a hacky way of avoiding that, agreed. Instead what we need is an estimate of the latency in the backend, and the time of computation for the wrapped function . If in the first call the latter takes longer than the latency by a factor of, say 2, then we store to cache and upon the next call, check these numbers again and retrieve from cache. Backends can define a default latency or, ideally, they can precompute it dynamically upon construction by averaging a few hundred dummy read/writes.

I think we should do this since it avoids a major potential performance issue and just takes a couple of calls to `time()`

Reply:
I understand and agree with your point. But I would rather leave it as is or even set it back to `0.3` for now because I still feel that caching may not actually be useful (We can determine that once we work on #459).

I will create a separate issue for your suggestion of determining the time_threshold dynamically.

Reply:
I believe 0 is not an option: it will be a performance hit and potentially create a bad network bottleneck. So 0.3 (although I think the dynamic thing is trivial to implement). Also, caching is necessary for at least two reasons:
* PermutationSampler
* Utility reuse for benchmarks (to avoid variance in the utility function).

Reply:
I will set it back to `0.3` for now. I agree that a simple implementation of the dynamic time_threshold estimation would be trivial but I would rather do it separately since I'm not convinced of the usefulness of caching, at least the way we are using it right now.

I would suggest thinking about utility reuse (e.g. for benchmarking) on its own and not try to tack it on caching.

Perhaps we could think of having an actual storage backend for the utility values by adding it directly to a Utility subclass (e.g. `StorageUtility`) or by replacing caching with key-value stores by an actual database as it is done in Optuna.